text/0089-scoped-api-keys.md

```diff
@@ -0,0 +1,128 @@
+- These `Scoped API keys` can't be less restrictive than a parent `API key` and can only be used for the search action with a predefined forced filter field.
+- These `Scoped API keys` are not stored and thus retrievable on MeiliSearch. This is why we highly advise setting an expiration time on that type of API key for security reasons.
+
+#### Solving Multi-Tenancy with `Scoped API Keys`
+
+![](https://i.imgur.com/J4jVe1n.png)
+
+Let's say that `Mark` is a developer for a SaaS platform. He would like to ensure that an end-user can only access his documents at search time. **His database contains many users and he hopes to have many more in the future.**
+
+When a user registers, the backend-side client code generates a `Scoped API Key` specifically for that end-user so he can only access his documents.
+
+A `filter` parameter is set to restrict the search for documents having a `user_id` attribute. This `filter` parameter is contained in the `Scoped API key` payload and cannot be modified during the search by the end-user making the request. `filter` can be made of any valid filters. e.g. `user_id = 10 and category = Romantic`
+
+This `Scoped API Key` is generated from a parent `API key` used to cipher the `Scoped API Key`. On the MeiliSearch side, this permits checking permissions and authorizing the `UserX` search request using this `Scoped API Key`.
+
+---
+
+#### Generating a `Scoped API Key`
+
+```javascript
+const scopedApiKeyRestrictions = {
+     "indexesPolicy": {
+         "products": {
+             "filter": "user_id = 1"
+         },
+         "reviews": {
+             "filter": "user_id = 1 AND published = true"
+         }
+     },
+    "expiresIn": 3600
+}
+
+export const generateScopedApiKey = () => {
+  return (parentApiKey: string, restrictions: scopedApiKeyRestrictions): string => {
+    //hash the parentApiKey and keep a prefix of 4 chars of the hashed parentApiKey
+    const prefixKey = crypto
+        .createHmac('sha256', parentApiKey)
+        .digest('hex')
+        .substr(0, 4);
+
+    //serialize restrictions (indexesPolicies object and expiresIn)
+    const queryParameters = serializeQueryParameters(restrictions);
+
+    //create the secured part (parentApiKey + queryParameters)
+    const securedKey = crypto
+      .createHmac('sha256', parentApiKey)
+      .update(queryParameters)
+      .digest('hex');
+
+    //return the generated `Scoped Api Key`
+    return Buffer.from(prefixKey + securedKey + queryParameters).toString('base64');
+  };
+};
+```
+
+##### scopedApiKeyRestrictions
+
+The format allows defining specific enforced search filters for accessible indexes (these indexes must be defined in the parent `Api Key` used to generate the `Scoped API key` and have the search action).
+
+If the user does not want to define specific filters for each index accessible to the end-user, he can use the `*` index wildcard rule.
```

Comment:
Some indexes may not be accessible?

Reply:
The accessible indexes are those defined by the parent `API Key` used (which must have the `search` action permission) to generate the `Scoped API Key`. This format here allows setting a filter on particular fields if there is no common field between documents coming from different accessible indexes or to force some particular `filter` rule for specific indexes.